File: postprocessing/plotgen3/plotgen.pl
#!/usr/bin/perl

###########################################################################
# Plotgen v3.0
#
#  
###########################################################################

package Plotgen;

use strict;

use CaseReader;
use OutputWriter;
use Cwd 'abs_path';
use Switch;
use Getopt::Std;
use File::Basename;
use File::Copy::Recursive qw(fcopy rcopy dircopy fmove rmove dirmove);
use File::Path;

# The location of MATLAB. External users probably will not need the 
# sudo -u matlabuser part.
my $MATLAB = "sudo -u matlabuser /usr/local/bin/matlab -nodisplay -nodesktop";

# Plotgen Version Number
my $VERSION = 3.0;

# Used to create a "random" output directory so multiple runs
# don't overwrite each other.
my $randInt = int(rand(999999999999999)) + 1;

# Image Conversion Settings
my $DPI = 300;
my $QUALITY = 100;

# Argument list

# If running in nightly mode, this value should be set to 1
my $nightly = 0;

# Specifies to overwrite a directory (Default: false)
my $overwrite = 0;
my @inputDirs;
my $output = "";
my $outputTemp = "";
my $plotLes = 0;
my $plotBest = 0;
my $plotDec = 0;

# Custom Color Definitions for "CLUBB_current" and "CLUBB_previous"
my $lt_blue = "[ 0.00, 0.63, 1.00 ]";
my $orange = "[ 0.94, 0.50, 0.16 ]";
# Arrays to cycle through when auto is set for lines
my @lineStyles = ("--", "-", "-.");
my @lineColors = ($orange, $lt_blue, "green", "red", "blue", "cyan", "yellow", "magenta");
my @lineWidths = (4, 3, 2.5, 2, 1.5, 1);

# Counters for automatic lines
my $lineStyleCounter = 0;
my $lineColorCounter = 0;
my $lineWidthCounter = 0;

my $outputIndex = "";

my $plotCount = 0;

# Do not keep permissions on file copies
$File::Copy::Recursive::KeepMode = 0;

# This is needed so we can restore the display after plotgen runs.
# We need to clear this environment variable before running, otherwise
# matlab will fail.
my $sessionType = $ENV{'DISPLAY'}; 
$SIG{INT} = "cleanup";
$SIG{CHLD} = "IGNORE";

main();

###############################################################################
# Plotgen 3 Main subroutine
###############################################################################
sub main()
{
	readArgs();

	# Ensure that Matlab can write to the temp output folder
	my $mode = 0777; chmod $mode, $outputTemp;

	print("Input Folders: @inputDirs\n");
	print("Output Folder: $output \n");

	$outputIndex = $outputTemp . "/index.html";

	# Fork to make MATLAB faster
	my $pid = fork();

	if($pid == 0) # Child
	{
		$ENV{'DISPLAY'} = '';

		system("mkfifo matlab_pipe");
		system("$MATLAB <> matlab_pipe");

		# Convert the eps files to jpq
		convertEps();

		OutputWriter->writeFooter($outputIndex);
	
		cleanup();
	
		print("Done! To display the plots, open: \n$output/index.html \nin your web browser.\n\n");
		print("Hit [Enter] to continue.");

		exit(0);
	}
	else # Parent
	{
		OutputWriter->writeHeader($outputIndex);
		runCases();

		# Quit matlab
		system("echo quit > matlab_pipe");
		system("rm matlab_pipe");

		print("\n");
		exit(0);
	}
}

###############################################################################
# Runs all of the .case file in the cases folder.
###############################################################################
sub runCases()
{
	# Counter used to place images
	my $count = 0;

	# Loop through each .case file so the case can be plotted
	my @cases = <cases/*.cas*>;
	foreach my $file (@cases) 
	{
		# Read the case file. If there is an error, exit.
		if (my $err = CaseReader->readCase($file))
		{
	    		print(STDERR $err, "\n");
	    		exit(1);
	    	}

		if(dataExists($CASE::CASE{'name'}) && ($CASE::CASE{'enabled'} ne 'false'))
		{
			# Print the case title to the HTML page
			OutputWriter->writeCaseTitle($outputIndex, $CASE::CASE{'headerText'});
	
			# Print any additional text/html specified
			if($nightly == 1) # If in nightly mode
			{
				my $nightlySubText = $CASE::CASE{'nightlyOutput'}{'subText'};
				my $nightlySubHtml = $CASE::CASE{'nightlyOutput'}{'subHtml'};
				
				# Check to see if there was any additional text specified. If there was,
				# write it to the HTML file.
	
				if($nightlySubText)
				{
					OutputWriter->writeSubHeader($outputIndex, $nightlySubText);
				}
				
				if($nightlySubHtml)
				{
					OutputWriter->writeSubHtml($outputIndex, $nightlySubHtml);
				}
			}
			else # If not in nightly mode
			{
				my $subText = $CASE::CASE{'additionalOutput'}{'subText'};
				my $subHtml = $CASE::CASE{'additionalOutput'}{'subHtml'};
				
				if($subText)
				{
					OutputWriter->writeSubHeader($outputIndex, $subText);
				}
				
				if($subHtml)
				{
					OutputWriter->writeSubHtml($outputIndex, $subHtml);
				}
	
			}
			
			# Check to see if this is a budget plot or standard plot
			if($CASE::CASE{'type'} eq "budget")
			{
				buildMatlabStringBudget($CASE::CASE, $count);
				
				# Convert the eps files to jpq
				convertEps($CASE::CASE{'name'} . "_" . $count . "_budget");

				# Add image file to HTML page
				placeImages($CASE::CASE{'name'} . "_" . $count . "_budget");
			}
			else
			{
				buildMatlabStringStd($CASE::CASE);
				
				# Add image file to HTML page
				placeImages($CASE::CASE{'name'}, $plotCount);
			}

			$count++;
		}
		else
		{
			#print("Not plotting $CASE::CASE{'headerText'}\n");
		}
	}
}

###############################################################################
# Converts all EPS files to JPG file
###############################################################################
sub convertEps()
{
	print("\nConverting images...\n");
	mkdir "$outputTemp/jpg" unless -d "$outputTemp/jpg";
	my @epsFiles = <$outputTemp/*eps>;
	foreach my $eps (@epsFiles)
	{
		my $filename = basename($eps);
		system("convert -density $DPI -quality $QUALITY -colorspace RGB $eps $outputTemp/jpg/$filename.jpg");

		unlink($eps);
	}
}

sub placeImages()
{
	my $caseName = shift(@_);
	my $numImages = shift(@_);

	OutputWriter->printDivCenter($outputIndex);

	for(my $x = 0; $x < $numImages; $x++)
	{
		OutputWriter->placeImage($outputIndex, "jpg/$caseName" . "_" . "$x.eps.jpg");
	}

	OutputWriter->printCloseDivCenter($outputIndex);
}

###############################################################################
# Generates argument lists needed for the Matlab part of Plotgen. This is for
# budget plots that will plot each input directory on a seperate plot.
###############################################################################
sub buildMatlabStringBudget()
{
	my $CASE = shift(@_);
	my $budgetNum = shift(@_);
	
	my $totPlotNum = 0;
	
	# Get Common Case information
	my $caseName =  $CASE::CASE{'name'} . "_" . $budgetNum . "_budget";
	my $startTime =  $CASE::CASE{'startTime'};
	my $endTime =  $CASE::CASE{'endTime'};
	my $startHeight =  $CASE::CASE{'startHeight'};
	my $endHeight =  $CASE::CASE{'endHeight'};
	
	# Get array of plots from case file
	my @plots;
	push(@plots, @{$CASE::CASE{'plots'}});

	my $units = $plots[0]{'axisLabel'};
	my $type = $plots[0]{'type'};
	
	# Loop through each input folder and create a plot for each folder
	for(my $plotCount = 0; $plotCount < @inputDirs; $plotCount++)
	{
		my $plotTitle = basename($inputDirs[$plotCount]);
		
		# Loop through each plot in the Case file
		for(my $plotNum = 0; $plotNum < @plots; $plotNum++)
		{
			# Counters for automatic lines
			$lineStyleCounter = 0;
			$lineColorCounter = 0;
			$lineWidthCounter = 0;
			my $matlabArgs = "\'$caseName\', \'$plotTitle\', $totPlotNum, \'$type\', $startTime, $endTime, $startHeight, $endHeight, \'$units\', $randInt";
			my $tempMatlabArgs = $matlabArgs;
		
			my @lines;
			push(@lines, @{$plots[$plotNum]{'lines'}});
			
			for(my $lineNum = 0; $lineNum < @lines; $lineNum++)
			{
				my $file = "$inputDirs[$plotCount]/$lines[$lineNum]{'filename'}";
				
				my $name = $lines[$lineNum]{'name'};
				my $expression = $lines[$lineNum]{'expression'};
				my $type = $lines[$lineNum]{'type'};

				if(-e $file)
				{
					my $lineWidth = $lineWidths[$lineWidthCounter];
					my $lineStyle = $lineStyles[$lineStyleCounter];
					my $lineColor = $lineColors[$lineColorCounter];
							
					$matlabArgs = "$matlabArgs, \'$file\', \'$expression\', \'$name\', $lineWidth, \'$lineStyle\', \'$lineColor\'";		

					incrementLineTypes();
				}
			}
			
			# Check to see if there are lines to be plotted:
			if($matlabArgs eq $tempMatlabArgs)
			{
				print(STDERR "No valid data available to plot.\n");
			}
			else
			{
				executeMatlab($matlabArgs);
				$totPlotNum ++;
			}
		}
	}
}

###############################################################################
# Generates the argument list needed for the Matlab part of Plotgen. This is
# for standard runs that plots variables from each input folder on the same
# plot.
###############################################################################
sub buildMatlabStringStd()
{
	my $CASE = shift(@_);

	$plotCount = 0;
	
	# Get Common Case information
	my $caseName =  $CASE::CASE{'name'};
	my $startTime =  $CASE::CASE{'startTime'};
	my $endTime =  $CASE::CASE{'endTime'};
	my $startHeight =  $CASE::CASE{'startHeight'};
	my $endHeight =  $CASE::CASE{'endHeight'};
	
	# Get array of plots from case file
	my @plots;
	push(@plots, @{$CASE::CASE{'plots'}});

	# Get plots from .case file
	for(my $count = 0; $count < @plots; $count++)
	{
		# Counters for automatic lines
		$lineStyleCounter = 0;
		$lineColorCounter = 0;
		$lineWidthCounter = 0;

		my $plotTitle = $plots[$count]{'plotTitle'};
		my $units = $plots[$count]{'axisLabel'};
		my $type = $plots[$count]{'type'};

		my $matlabArgs = "\'$caseName\', \'$plotTitle\', $count, \'$type\', $startTime, $endTime, $startHeight, $endHeight, \'$units\', $randInt";
		my $tempMatlabArgs = $matlabArgs;

		my @lines;
		push(@lines, @{$plots[$count]{'lines'}});

		for(my $lineNum = 0; $lineNum < @lines; $lineNum++)
		{
			my $name = $lines[$lineNum]{'name'};
			my $expression = $lines[$lineNum]{'expression'};

			my $type = $lines[$lineNum]{'type'};

			if(($type eq "les" && $plotLes == 1) || ($type eq "dec17" && $plotDec) || ($type eq "bestever" && $plotBest))
			{
				my $file = "$lines[$lineNum]{'filename'}";
				if(-e $file)
				{
					my $title;

					if($type eq "les")
					{
						$title = "LES";
					}
					elsif($type eq "dec17")
					{
						$title = "HOC 12/17/2005";
					}
					elsif($type eq "bestever")
					{
						$title = "HOC \"best-ever\"";
					}

					my $lineWidth = $lines[$lineNum]{'lineWidth'};
					my $lineStyle = $lines[$lineNum]{'lineType'};
					my $lineColor = $lines[$lineNum]{'lineColor'};
	
					$matlabArgs = "$matlabArgs, \'$file\', \'$expression\', \'$title\', $lineWidth, \'$lineStyle\', \'$lineColor\'";
				}
			}
			elsif($type eq "auto")
			{
				foreach (@inputDirs)
				{
					my $file = "$_/$lines[$lineNum]{'filename'}";

					if(-e $file)
					{
						my $title = basename($_);
						
						my $lineWidth = $lineWidths[$lineWidthCounter];
						my $lineStyle = $lineStyles[$lineStyleCounter];
						my $lineColor = $lineColors[$lineColorCounter];
						
						$matlabArgs = "$matlabArgs, \'$file\', \'$expression\', \'$title\', $lineWidth, \'$lineStyle\', \'$lineColor\'";

						incrementLineTypes();
					}
				}
			}
		}

		# Check to see if there are lines to be plotted:
		if($matlabArgs eq $tempMatlabArgs)
		{
			#print(STDERR "No valid data available to plot.\n");
		}
		else
		{
			$plotCount++;
			executeMatlab($matlabArgs);	
		}
	}
}

sub executeMatlab()
{
	my $matlabArgs = shift(@_);

	#my $args = "echo \"quit\" | $MATLAB -nodisplay -nodesktop -r PlotCreator\"($matlabArgs)\"";
	my $args = "PlotCreator\"($matlabArgs)\"";

	#print $args;

	system("echo $args > matlab_pipe");
}

###############################################################################
# Increments the counters for the automatic line types
###############################################################################
sub incrementLineTypes()
{
	# Increment counters
	if($lineColorCounter + 1 >= @lineColors)
	{
		$lineColorCounter = 0;

		if($lineStyleCounter + 1 >= @lineStyles)
		{
			$lineStyleCounter = 0;

			if($lineWidthCounter + 1 >= @lineWidths)
			{
				$lineWidthCounter = 0;
			}
			else
			{
				$lineWidthCounter ++;
			}
		}
		else
		{
			$lineStyleCounter ++;

			if($lineWidthCounter + 1 >= @lineWidths)
			{
				$lineWidthCounter = 0;
			}
			else
			{
				$lineWidthCounter ++;
			}
		}
	}
	else
	{
		$lineColorCounter++;

		if($lineStyleCounter + 1 >= @lineStyles)
		{
			$lineStyleCounter = 0;
	
			if($lineWidthCounter + 1 >= @lineWidths)
			{
				$lineWidthCounter = 0;
			}
			else
			{
				$lineWidthCounter ++;
			}
		}
		else
		{
			$lineStyleCounter ++;

			if($lineWidthCounter + 1 >= @lineWidths)
			{
				$lineWidthCounter = 0;
			}
			else
			{
				$lineWidthCounter ++;
			}
		}
	}
}

sub cleanup()
{
	# Copy temp. output folder to actual output location and remove the temp. folder
	dircopy($outputTemp, $output);
	rmtree($outputTemp);

	$ENV{'DISPLAY'} = $sessionType;
}

###############################################################################
# Checks all input directories to see if one of them contains the current case.
# Will return true if at least on input folder contains data, otherwise false.
# #############################################################################
sub dataExists()
{
	my $dataFile = shift(@_);
	my $retValue = 0;
	
	foreach (@inputDirs)
	{
		my @files = <$_/$dataFile*>;
		if(@files)
		{
			$retValue = 1;
		}
	}

	return $retValue;
}

###############################################################################
# Reads any arguments passed in.
###############################################################################
sub readArgs()
{
	my $numArgs = $#ARGV + 1;

	if($numArgs == 0)
	{
		main::HELP_MESSAGE();
	}

	my %option = ();
	getopts("rlbdanh?", \%option);

	if ($option{r})
	{
		$overwrite = 1;
	}

	if ($option{l})
	{
		$plotLes = 1;
	}

	if ($option{b})
	{
		$plotBest = 1;
	}

	if ($option{d})
	{
		$plotDec = 1;
	}

	if ($option{a})
	{
		$plotLes = 1;
		$plotBest = 1;
		$plotDec = 1;
	}

	if ($option{n})
	{
		$nightly = 1;
	}

	if ($option{h})
	{
		main::HELP_MESSAGE();
	}

	my $currentCount = 0;

	# Reset the number of arguments
	$numArgs = $#ARGV + 1;

	# Parse any additional arguments
	foreach (@ARGV)
	{
		# If the argument does not start with '-' and if $output was not set
		if(!$output)
		{
			my $currentDir = abs_path($_);

			if((($numArgs - 1) - $currentCount) > 0)
			{
				if(-d $currentDir)
				{
					push(@inputDirs, $currentDir);
				}
				else
				{
					print("The input folder: $currentDir does not exist.\n");
					exit(1);
				}
			}
			else
			{
				if($_ =~ m/\/$/)
				{
					$currentDir = abs_path(substr($_, 0, length($_) - 1));
				}

				$output = $currentDir;
				$outputTemp = "/tmp/output" . "_" . "$randInt";
			}
			
			$currentCount++;
		}
	}

	if(@inputDirs == 0)
	{
		main::HELP_MESSAGE();
	}

	# Finally, check to see if the output folder exists. If it does, and
	# '-r' was not passed in, exit. Otherwise, create it.
	if(-d $output && $overwrite == 0)
	{
		print("Output folder already exists. To overwrite, use the -r option.\n");
		exit(1);
	}
	else
	{
		rmtree($output);
		mkdir $output;
		mkdir $outputTemp;
	}
}

###############################################################################
# Prints the help message
###############################################################################
sub main::HELP_MESSAGE()
{
	print("Usage: plotgen [OPTION]... INPUT... OUTPUT\n");
	print("  -r\tIf the output folder already exists, replace the contents\n");	
	print("  -l\tPlot LES data for comparison.\n");	
	print("  -b\tPlot Best Ever data for comparison.\n");	
	print("  -d\tPlot December data for comparison.\n");	
	print("  -a\tSame as -lbd. Plots LES, Best Ever, and December data for comparison.\n");
	print("  -n\tRuns in nightly mode.\n");
	print("  -h\tPrints this help message.\n");
	exit(0);
}

###############################################################################
# Prints the version number of Plotgen.
###############################################################################
sub main::VERSION_MESSAGE()
{
	print("Plotgen version $VERSION, Copyright (c) 2009 Larson Group.\n");
}

File: postprocessing/plotgen/plotgen.py
#!/usr/bin/python
###########################################################################
# Plotgen v4.0a
#
# Plotgen4 is written to be identical from a user standpoint to plotgen3, 
# except not requiring the use of Matlab.
# Documentation for plotgen3 is available here until plotgen4 docs are complete:
# http://larson-group.com/twiki/bin/view.pl/Documentation/CarsonDoc/Plotgen3
###########################################################################
#TODO add imports
import sys, os, OutputWriter, importlib

#TODO document these default variables even though they're kinda self-explanatory.

# Location of Script
dirPrefix = os.path.dirname(os.path.realpath(__file__))

# Plotgen Version
VERSION	= 4.0

# Run in Nightly mode: 		Default False
nightly = False

# Difference run: 		Default False
diffRun = False

# Plotgen mode dictates where to find .case files.
#   plotgen (Default)
#   splotgen
#   wrfgen
#   camgen
#   gfdlgen
plotgenMode = "plotgen"

# Data file type:  grads or netcdf
dataFileType = "grads"

# Budget plotting mode:		Default false
plotBudgets = False

# Morrison budget plot mode:	Default false
plotMorrBudgets = False

# Output as .maff file:		Default false
outputAsMaff = False

# Display legend on plots:	Default true
displayLegend = True

# Allow output folder to exist: Default false
overwrite = False

# Ensemble run flag:		Default false
ensembleTuner = False
thinLines = False

# Line style, etc, arrays
lineStyles = ["-", "--", "-.", "-"]
lineColors = [
    "[ 0.000, 0.000, 1.000 ]",    # blue
    "[ 0.000, 1.000, 0.000 ]",    # green
    "[ 1.000, 0.000, 1.000 ]",    # magenta
    "[ 0.471, 0.471, 0.471 ]",    # gray
    "[ 0.969, 0.506, 0.749 ]",    # pink
    "[ 1.000, 0.498, 0.000 ]",    # orange
    "[ 0.651, 0.337, 0.157 ]",    # brown
    "[ 0.894, 0.102, 0.110 ]",    # red
    "[ 1.000, 1.000, 0.200 ]"     # yellow
]
lineWidths = [4.5, 3, 2.5, 1.5]
lineStyleCounter = 0
lineColorCounter = 0
lineWidthCounter = 0

# Set width for budget cases.
lineWidthsBudget = 3

outputIndex = ""

# Plot benchmark lines:		Default False
plotAll = False
plotLes = False
plotBest = False
plotDec = False

# HTML page names
navigationPage = "navigation.html"
indexPage = "index.html"

plotcount = 0
consoleOutput = dirPrefix + "/console_output.py"
inputDirs = []
output = ""
outputTemp = ""
casesExecuted = []
DPI = 300
QUALITY = 100
caseCount = 0

###############################################################################
#  Main function
###############################################################################
def main():
##TODO do stuff
    sys.exit(0)

#CASE.CASE

###############################################################################
#  Runs through all cases in case folder.
###############################################################################
def runCases():
    imgNumber = 0
    casePath = getCasePath()
    #This portion will check if we are in nightly mode.  If not, we'll ignore
    #any case files with _nightly.
    if nightly:
        cases = [f for f in os.listdir(casePath) if f.endswith(".case")]
    else:
        cases = [f for f in os.listdir(casePath) 
                 if f.endswith(".case") and "_nightly" not in f]
    for filename in cases:
        runCase = true
        CASE = imp.load_source('CASE',casePath + filename)  ## Python doesn't like import filename
        if CASE.CASE['name'] in casesExecuted: runCase = false

        if runCase and dataExists() and CASE.CASE['enabled'] != false:
            casesExecuted.append(CASE.CASE['name'])
            if (CASE.CASE['type'] == "standard") or
               (CASE.CASE['type'] == "budget" and plotBudgets) or
               (CASE.CASE['type'] == "morrbudget" and plotMorrBudgets):
                OutputWriter.writeCaseTitle(outputIndex,case['headerText'])
                OutputWriter.writeNavePageCase(navigationPage, 
                                               CASE.CASE['name'], CASE.CASE['headerText'])
                # Output additional text from casefile
                if nightly:
                    if CASE.CASE['nightlyOutput']['subText']: 
                        OutputWriter.writeSubHeader(outputIndex, CASE.CASE['nightlyOutput']['subText'])
                    if CASE.CASE['nightlyOutput']['subHtml']:
                        OutputWriter.writeSubHtml(outputIndex, CASE.CASE['nightlyOutput']['subHtml'])
                else:
                    if CASE.CASE['type'] == "morrbudget":  ##This was in the original code
                        OutputWriter.writeMorrBudgetSubHeader(
                            outputIndex, CASE.CASE['additionalOutput']['subText'])
                    if CASE.CASE['additionalOutput']['subText']: 
                        OutputWriter.writeSubHeader(outputIndex, CASE.CASE['additionalOutput']['subText'])
                    if CASE.CASE['additionalOutput']['subHtml']:
                        OutputWriter.writeSubHtml(outputIndex, CASE.CASE['additionalOutput']['subHtml'])
            

                

###############################################################################
# Executes the PlotCreator.
# Arguments:
#   executeMatlab(String matlabArgs)
#     - matlabArgs: The arguments to be passed to the plot creator
###############################################################################
def executePlot():
#TODO maybe objectify this?  Each plot perhaps will be its own object
    sys.exit(1)

###############################################################################
# Changes to the next line style, width, and color.
# Arguments:
#   None.
###############################################################################
def incrementLineTypes():
    lineColorCounter = ((lineColorCounter + 1) % len(lineColors))
    lineStyleCounter = ((lineStyleCounter + 1) % len(lineStyles))
    lineWidthCounter = ((lineWidthCounter + 1) % len(lineWidths))
### This should work fine, if not, below will work when testing.
#    if lineColorCounter >= len(lineColors): lineColorCounter = 0
#    else: lineColorCounter=lineColorCounter+1
#    if lineStyleCounter >= len(lineStyles): lineStyleCounter = 0
#    else: lineStyleCounter=lineStyleCounter+1
#    if lineWidthCounter >= len(lineWidths): lineWidthCounter = 0
#    else: lineWidthCounter=lineWidthCounter+1

###############################################################################
# Does necessary cleanup code.
# Arguments:
#   None.
###############################################################################
def cleanup():
# TODO clean up all temp files and other cruft.
    sys.exit(1)

###############################################################################
# Checks all input directories to see if one of them contains the current case.
# Will return true if at least on input folder contains data, otherwise false.
# Arguments:
#   dataExists(Hash CASE)
#     - CASE: The case file to check
###############################################################################
def dataExists():
# TODO take in case file, pull list of plots, iterate through all and find datafile names
# check input folders for data files and if the right data file is there for the case
# If so, return 1 (which then runs the case plots elsewhere) else return 0.
# there may be a quicker way for this to work, eventually.
# To do this, I'll need to translate the case files over.
    return 0

###############################################################################
#  Casefile path generator function
###############################################################################
def getCasePath():
    casePath = dirPrefix + "cases"
    if plotgenMode == "plotgen":
        return casePath + "/clubb"
    elif plotgenMode == "splotgen":
        return casePath + "/sam_clubb"
    elif plotgenMode == "wrfgen":
        return casePath + "/wrf"
    elif plotgenMode == "camgen":
        return casePath + "/cam"
    elif plotgenMode == "gfdlgen":
        return casePath + "/gfdl"
    else: sys.exit("Plotgen run mode unknown, please use plotgen.py -h for help")
# This error should not be possible unless code was changed.

###############################################################################
#  User is confused.  Possibly drunk.  Print a help message.
###############################################################################
def HELP_MESSAGE():
    print("Usage: plotgen [OPTION]... INPUT... OUTPUT\n")
    print("  -c\tPlot CLUBB cases [DEFAULT] (equiv to plotgen)\n")
    print("  -s\tPlot SAM_CLUBB cases (equiv to splotgen)\n")
    print("  -w\tPlot WRF_CLUBB cases\n")
    print("  -cam\tPlot CAM cases\n")
    print("  -gfdl\tPlot GFDL cases\n")
    print("  -r\tIf the output folder already exists, replace the contents\n")
    print("  -l\tPlot LES data for comparison.\n")
    print("  -b\tPlot HOC Best Ever data for comparison.\n")
    print("  -d\tPlot HOC 12/17/2005 data for comparison.\n")
    print("  -a\tSame as -lbd. Plots LES, Best Ever, and 12/17/2005 " + 
          "data for comparison.\n")
    print("  -n\tRuns in nightly mode.\n")
    print("  -q\tOutputs high quality images (does not auto scale).\n")
    print("  -e\tDoes not delete EPS images after conversion.\n")
    print("  -m\tOutputs plots compressed inside a .maff directory.\n")
    print("  -g\tUses GrADS data files. [DEFAULT]\n")
    print("  -t\tUses NetCDF data files.\n")
    print("  -thin\tUses thin solid lines\n")
    print("  -dfrnce\tPerforms a 'difference' plot of two output folders\n")
    print("  -nolegend\tPlot without legends\n")
    print("  -ensemble\tUsed for plotting ensemble tuner runs\n")
    print("  -bu\tUsed to plot standard budget plots\n")
    print("  -bumorr\tUsed to plot Morrison budget plots\n")
    print("  -h\tPrints this help message.\n")
    sys.exit(0)

###############################################################################
#  Parse arguments
###############################################################################
def parseArgs(argList):
    numArgs = 0
    if len(argList) == 0:
        helpMessage()
    for arg in argList:
        if arg[0] == "-":
            numArgs += 1
            arg = arg[1:]  
            for curarg in arg:
                if   curarg == 'm': outputAsMaff = True
                elif curarg == 'e': keepEps = True
                elif curarg == 'a':
                    plotLes = True
                    plotBest = True
                    plotDec = True
                elif curarg == 'r': overwrite = True
                elif curarg == 'b': plotBest = True
                elif curarg == 'd': plotDec = True
                elif curarg == 'l': plotLes = True
                elif curarg == 'n': nightly = True
                elif curarg == 'q': highQuality = True
                elif curarg == 'g': dataFileType = "grads"
                elif curarg == 't': dataFileType = "netcdf"
                elif curarg == 'thin': thinLines = True
                elif curarg == 'dfrnce': diffRun = True
                elif curarg == 'ensemble': ensembleTuner = True
                elif curarg == 'nolegend': displayLegend = False
                elif curarg == 's': plotgenMode = "splotgen"
                elif curarg == 'c': plotgenMode = "plotgen"
                elif curarg == 'w': plotgenMode = "wrfgen"
                elif curarg == 'cam': plotgenMode = "camgen"
                elif curarg == 'gfdl': plotgenMode = "gfdlgen"
                elif curarg == 'bu': plotBudgets = True
                elif curarg == 'bumorr': plotMorrBudgets = True
                elif curarg == 'h': HELP_MESSAGE()
                else: HELP_MESSAGE()

    if keepEps and outputAsMaff:
        sys.exit("Argument conflict: Please do not simultaneously" + 
                 "choose to save .eps files (-e) and create a .maff" +
                 "file (-m)\n")
    if len(argList)-numArgs < 2:
        HELP_MESSAGE()
    inputDirs = []
    for dir in argList[numArgs:len(argList)-1]:
        if os.path.isdir(dir):
            inputDirs.append(os.path.join(dir)) 
        else:
            print("Input folder " + dir + " does not exist.")
            sys.exit(1)
    outputDir = argList[len(argList)-1]
    if os.path.isdir(outputDir):
        outputDir = os.path.join(outputDir)
        if overwrite:
            try:
                import shutil
                shutil.rmtree(outputDir)
                os.mkdir(outputDir)
            except Exception, e:
                print(e)
                sys.exit(1)
        else:
            print("Output folder alread exists.  Please use -r to allow overwrite")
            sys.exit(1)
    else:
        try:
            os.mkdir(outputDir)
        except Exception, e:
            print e
            sys.exit(1)

###############################################################################
#  Print the version number of Plotgen
###############################################################################
def VERSION_MESSAGE():
    print('Plotgen version $VERSION, Copyright (C) 2013 Larson Group.\n')


# Now that 
main()
